### PR TITLE
pkg/plugin/v3: add plugin `v3-alpha`

### PR DESCRIPTION
**Pull request by estroz**
A new plugin version `v3-alpha` will contain all new plugin/scaffold changes/features. Since plugin version `v2` is tested with project version 2, we probably should just test the new plugin with project version `3-alpha`. This won't necessarily be the case going forward, ex. once `4-alpha` is created we should generate separate testdata.

/cc @DirectXMan12 @camilamacedo86 @joelanford @mengqiy

### CODE REVIEW

**Comment by estroz** [comment]
A thought on bumping plugin version stages, ex. `alpha` to `beta` and `beta` to "" (stable): we should probably:
1. Just change the plugin's stage string instead of creating a new package.
1. Build a mechanism into `pkg/cli` that auto-upgrades an `alpha` or `beta` to the next stage if available. This should never be an issue because the next stage should be no different from the previous one, just more stable.

Nothing needs to be done in this PR regarding ^.

**Comment by estroz** [comment]
We may want to consider adding e2e test for plugin version `v3-alpha` too. Will likely follow up with these.

File: pkg/plugin/v3/api.go [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+	"sigs.k8s.io/kubebuilder/pkg/plugin"
+	"sigs.k8s.io/kubebuilder/pkg/plugin/internal/util"
+	"sigs.k8s.io/kubebuilder/pkg/plugin/scaffold"
+	"sigs.k8s.io/kubebuilder/pkg/plugin/v3/scaffolds"
+	"sigs.k8s.io/kubebuilder/plugins/addon"
+)
+
+// (used only to gen api with --pattern=addon)
+// KbDeclarativePatternVersion is the sigs.k8s.io/kubebuilder-declarative-pattern version
+const KbDeclarativePatternVersion = "v0.0.0-20200522144838-848d48e5b073"
```
Comment (DirectXMan12) [inline pkg/plugin/v3/api.go:43]: do we need this?  I'm assuming going forward we just want `--plugin declarative` or something

File: pkg/plugin/v3/api.go [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+		"if set, generate the resource without prompting the user")
+	p.resourceFlag = fs.Lookup("resource")
+	fs.BoolVar(&p.doController, "controller", true,
+		"if set, generate the controller without prompting the user")
+	p.controllerFlag = fs.Lookup("controller")
+
+	if os.Getenv("KUBEBUILDER_ENABLE_PLUGINS") != "" {
+		fs.StringVar(&p.pattern, "pattern", "",
+			"generates an API following an extension pattern (addon)")
+	}
```
Comment (DirectXMan12) [inline pkg/plugin/v3/api.go:114]: ditto here

File: pkg/plugin/v3/api.go [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+
+func (p *createAPIPlugin) Validate() error {
+	if err := p.resource.Validate(); err != nil {
+		return err
+	}
+
+	reader := bufio.NewReader(os.Stdin)
+	if !p.resourceFlag.Changed {
+		fmt.Println("Create Resource [y/n]")
+		p.doResource = util.YesNo(reader)
```
Comment (DirectXMan12) [inline pkg/plugin/v3/api.go:141]: at some point we should re-evaluate the yes/no stuff.  IIRC we had wanted to do it for v2, but didn't want to break compat

File: pkg/plugin/v3/api.go [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+	if p.doResource {
+		// Check that resource doesn't exist or flag force was set
+		if !p.force && p.config.HasResource(p.resource.GVK()) {
+			return errors.New("API resource already exists")
+		}
+
+		// Check that the provided group can be added to the project
+		if !p.config.MultiGroup && len(p.config.Resources) != 0 && !p.config.HasGroup(p.resource.Group) {
+			return fmt.Errorf("multiple groups are not allowed by default, to enable multi-group visit %s",
+				"kubebuilder.io/migration/multi-group.html")
```
Comment (DirectXMan12) [inline pkg/plugin/v3/api.go:158]: why is this a `%s`?

File: pkg/plugin/v3/api.go [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+}
+
+func (p *createAPIPlugin) PostScaffold() error {
+	// Load the requested plugins
+	switch strings.ToLower(p.pattern) {
+	case "":
+		// Default pattern
+	case "addon":
+		// Ensure that we are pinning sigs.k8s.io/kubebuilder-declarative-pattern version
+		err := util.RunCmd("Get controller runtime", "go", "get",
```
Comment (DirectXMan12) [inline pkg/plugin/v3/api.go:195]: ??
this isn't getting CR, and shouldn't it be using the constant from above anyway?

File: pkg/plugin/v3/init.go [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+	if err != nil {
+		return err
+	}
+
+	err = util.RunCmd("Update go.mod", "go", "mod", "tidy")
+	if err != nil {
+		return err
+	}
+
+	err = util.RunCmd("Running make", "make")
```
Comment (DirectXMan12) [inline pkg/plugin/v3/init.go:158]: shouldn't this be conditional on the make flag defined in the other file?

File: pkg/plugin/v3/scaffolds/api.go [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+		model.WithResource(s.resource),
+	)
+}
+
+func (s *apiScaffolder) scaffold() error {
+	if s.doResource {
+		s.config.AddResource(s.resource.GVK())
+
+		if err := machinery.NewScaffold(s.plugins...).Execute(
+			s.newUniverse(),
```
Comment (DirectXMan12) [inline pkg/plugin/v3/scaffolds/api.go:85]: shouldn't we be passing the universe between scaffold calls here?

File: pkg/plugin/v3/scaffolds/init.go [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+	return s.scaffold()
+}
+
+func (s *initScaffolder) scaffold() error {
+	bpFile := &templates.Boilerplate{}
+	bpFile.Path = s.boilerplatePath
+	bpFile.License = s.license
+	bpFile.Owner = s.owner
+	if err := machinery.NewScaffold().Execute(
+		s.newUniverse(""),
```
Comment (DirectXMan12) [inline pkg/plugin/v3/scaffolds/init.go:85]: ditto here and elsewhere re: re-using the universe

**Comment by DirectXMan12** [comment]
> If the answer to 1 is no (alpha implies the plugin is unstable), then the previous stable plugin version should be the default (v2 in this case). If yes, then v3-alpha is fine.

Per the meeting, IIRC we decided that the answer was "no", and that the default scaffolding for the v3 project version was going to be v2 till v3 stable was ready

File: pkg/plugin/v3/api.go (hunk 5 above)
Comment (estroz) [reply pkg/plugin/v3/api.go:195]: This is a typo, it should be "Get kubebuilder-declarative-pattern dependency". I'll change this text for now, but also add a comment to handle  the issue this tries to fix.

File: pkg/plugin/v3/api.go (hunk 1 above)
Comment (estroz) [reply pkg/plugin/v3/api.go:43]: I think this attempts to fix a CI issue where the testdata scaffolded with addons enabled will pull the latest kubebuilder-declarative-pattern commit, which messes up `git diff`.

We should probably scaffold a `tools.go` containing a pinned version of this. I could also see kustomize and controller-tools versions being pinned this way, which are currently dealt with in the Makefile.

File: pkg/plugin/v3/init.go (hunk 6 above)
Comment (estroz) [reply pkg/plugin/v3/init.go:158]: Currently there's no `init --make` flag unlike `create api`. Perhaps we should add one.

**Comment by estroz** [comment]
@DirectXMan12 fixed what was easy/reasonable and added TODO's for the rest.

File: pkg/plugin/v3/api.go [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+		// Default pattern
+	case "addon":
+		// Ensure that we are pinning sigs.k8s.io/kubebuilder-declarative-pattern version
+		// TODO: either find a better way to inject this version (ex. tools.go).
+		err := util.RunCmd("Get kubebuilder-declarative-pattern dependency", "go", "get",
+			"sigs.k8s.io/kubebuilder-declarative-pattern@"+KbDeclarativePatternVersion)
+		if err != nil {
+			return err
+		}
+	default:
```
Comment (camilamacedo86) [inline pkg/plugin/v3/api.go:205]: Why change the impl that was done in https://github.com/kubernetes-sigs/kubebuilder/pull/1527/files?

File: pkg/plugin/v3/api.go (hunk 1 above)
Comment (camilamacedo86) [reply pkg/plugin/v3/api.go:44]: Hi @estroz why change the impl that was done in https://github.com/kubernetes-sigs/kubebuilder/pull/1527/files?

File: pkg/plugin/v3/init.go (hunk 6 above)
Comment (camilamacedo86) [reply pkg/plugin/v3/init.go:159]: Should we not here just replicate so the code to v3? IMO we should not change any impl in this pr just add the v3-alpha plugin and then we can check each case/scenario in new pr's.

Comment (estroz) [reply pkg/plugin/v3/init.go:159]: This won't be changed here, just added a TODO.

File: pkg/plugin/v3/api.go (hunk 1 above)
Comment (estroz) [reply pkg/plugin/v3/api.go:44]: This won't be changed here, just added a TODO.

Comment (camilamacedo86) [reply pkg/plugin/v3/api.go:44]: OK. Yep, we have an issue tracked for we create a plugin for that. See: https://github.com/kubernetes-sigs/kubebuilder/issues/1543

File: pkg/plugin/v3/scaffolds/api.go (hunk 7 above)
Comment (camilamacedo86) [reply pkg/plugin/v3/scaffolds/api.go:86]: I think that it also can be a TODO @estroz ?

File: pkg/plugin/v3/scaffolds/init.go (hunk 8 above)
Comment (camilamacedo86) [reply pkg/plugin/v3/scaffolds/init.go:86]: I think that it also can be a TODO @estroz ?

File: pkg/plugin/v3/init.go [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+- a go.mod with project dependencies
+- a Kustomization.yaml for customizating manifests
+- a Patch file for customizing image for manager manifests
+- a Patch file for enabling prometheus metrics
+- a cmd/manager/main.go to run
+
+project will prompt the user to run 'dep ensure' after writing the project files.
+`
+	ctx.Examples = fmt.Sprintf(`  # Scaffold a project using the apache2 license with "The Kubernetes authors" as owners
+  %s init --project-version=2 --domain example.org --license apache2 --owner "The Kubernetes authors"
```
Comment (camilamacedo86) [inline pkg/plugin/v3/init.go:71]:
```suggestion
  %s init --project-version=3-alpha --domain example.org --license apache2 --owner "The Kubernetes authors"
```

File: pkg/plugin/v3/init.go [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+	skipGoVersionCheck bool
+}
+
+var (
+	_ plugin.Init        = &initPlugin{}
+	_ cmdutil.RunOptions = &initPlugin{}
+)
+
+func (p *initPlugin) UpdateContext(ctx *plugin.Context) {
+	ctx.Description = `Initialize a new project including vendor/ directory and Go package directories.
```
Comment (camilamacedo86) [inline pkg/plugin/v3/init.go:56]:
```suggestion
	ctx.Description = `Initialize a new project including Go package directories.
```
I should be removed already.

File: pkg/plugin/v3/init.go (hunk 10 above)
Comment (estroz) [reply pkg/plugin/v3/init.go:71]: Follow-up

File: pkg/plugin/v3/init.go (hunk 11 above)
Comment (estroz) [reply pkg/plugin/v3/init.go:56]: Follow-up

**Pull request merged**